File: chp2_recursive_divide_conquer/divide_n.py
def divide_n(n, m):
    if n < 1 or m < 1:
        return 0
    elif n == 1 or m == 1:
        return 1
    elif n < m:
        return divide_n(n, n)
    elif n == m:
        return divide_n(n, m-1) + 1
    else:
        return divide_n(n, m-1) + divide_n(n-m, m)


n = int(input('Input a number: '))
print(f'The divide of {n} has {divide_n(n,n)} methods ')

File: chp2_recursive_divide_conquer/divide_n.cpp
#include <iostream>
using namespace std;

int q(int n, int m) {
  if (n < 1 || m < 1)
    return 0;
  else if (n == 1 || m == 1) {
    return 1;
  } else if (n < m) {
    return q(n, n);
  } else if (n == m) {
    return q(n, m - 1) + 1;
  } else {
    return q(n, m - 1) + q(n - m, m);
  }
}

int main(void) {
  cout << "Input a number:\n";
  int n;
  cin >> n;
  cout << "The divide of the number has " << q(n,n) <<" methods.";
  return 0;
}
